validation/src/validation_service/mod.rs

```diff
@@ -384,98 +461,315 @@ impl<C, N, P, SP> ParachainValidationInstances<C, N, P, SP> where
 	fn retain<F: FnMut(&Hash) -> bool>(&mut self, mut pred: F) {
 		self.live_instances.retain(|k, _| pred(k))
 	}
+}
 
-	// launch parachain work asynchronously.
-	fn launch_work(
-		&self,
-		relay_parent: Hash,
-		validation_para: ParaId,
-		build_router: N::BuildTableRouter,
-		max_block_data_size: Option<u64>,
-		n_validators: usize,
-		local_id: ValidatorIndex,
-	) {
-		let (collators, client) = (self.collators.clone(), self.client.clone());
-		let availability_store = self.availability_store.clone();
+// launch parachain work asynchronously.
+async fn launch_work<Client>(
+	collation_fetch: impl CollationFetch,
+	client: Arc<Client>,
+	availability_store: AvailabilityStore,
+	relay_parent: Hash,
+	validation_para: ParaId,
+	router: impl TableRouter,
+	max_block_data_size: Option<u64>,
+	n_validators: usize,
+	local_id: ValidatorIndex,
+) where
+	Client: ProvideRuntimeApi<Block> + Send + Sync + 'static,
+	Client::Api: ParachainHost<Block, Error = sp_blockchain::Error>,
+{
+	// fetch a local collation from connected collators.
+	let (collation_info, full_output) = match collation_fetch.collation_fetch(
```

Comment:
One simplification would be to accept the `collation_fetch` future as an argument rather than the `CollationFetch` trait. then this function doesn't need all those parameters